play.pokemonshowdown.com/src/panel-chat.tsx

```diff
@@ -669,6 +701,107 @@ export class ChatTextEntry extends preact.Component<{
 		}
 		return false;
 	}
+	/**
+	 * Taken from the old client — just tweaked a bit to:
+     * - Work with <pre> instead of <textarea>
+     * - Fit the way `room` now handles user info
+	 * TODO - add support for commands tabcomplete
+     */
+	handleTabComplete(reverse: boolean) {
+		// Don't tab complete at the start of the text box.
+		let idx = this.getSelection().start;
+		if (idx === 0) return false;
+		let users = (this.props.room as ChatRoom)?.users || (PS.rooms['lobby'] ? (PS.rooms['lobby'] as ChatRoom).users : {});
+		let text = this.miniedit?.getValue() || '';
+		let prefix = text.substr(0, idx);
```

Comment:
```suggestion
		let prefix = text.slice(0, idx);
```

`substr` is deprecated; I've been mostly using `slice` in new code.